tests/semantic_analyser.cpp

```diff
@@ -1067,6 +1068,62 @@ TEST(semantic_analyser, call_len)
   test("kprobe:f { $x = 0; len($x); }", 1);
 }
 
+TEST(semantic_analyser, call_has_key)
+{
+  test("kprobe:f { @x[1] = 0; if (has_key(@x, 1)) {} }");
+  test("kprobe:f { @x[1, 2] = 0; if (has_key(@x, (3, 4))) {} }");
+  test("kprobe:f { @x[1, (int8)2] = 0; if (has_key(@x, (3, 4))) {} }");
+  test(R"(kprobe:f { @x[1, "hi"] = 0; if (has_key(@x, (2, "bye"))) {} })");
+  test(
+      R"(kprobe:f { @x[1, "hi"] = 0; if (has_key(@x, (2, "longerstr"))) {} })");
+  test(
+      R"(kprobe:f { @x[1, "longerstr"] = 0; if (has_key(@x, (2, "hi"))) {} })");
+  test("kprobe:f { @x[1, 2] = 0; $a = (3, 4); if (has_key(@x, $a)) {} }");
+  test("kprobe:f { @x[1, 2] = 0; @a = (3, 4); if (has_key(@x, @a)) {} }");
+
+  test_error("kprobe:f { @x[1] = 1;  if (has_key(@x)) {} }",
+             R"(
+stdin:1:27-39: ERROR: has_key() requires at least 2 arguments (1 provided)
+kprobe:f { @x[1] = 1;  if (has_key(@x)) {} }
+                          ~~~~~~~~~~~~
+)");
+
+  test_error("kprobe:f { @x[1] = 1;  if (has_key(@x[1], 1)) {} }",
+             R"(
+stdin:1:27-41: ERROR: has_key() expects the first argument to be a map. Not a map value expression.
+kprobe:f { @x[1] = 1;  if (has_key(@x[1], 1)) {} }
+                          ~~~~~~~~~~~~~~
+)");
+
+  test_error("kprobe:f { @x = 1;  if (has_key(@x, 1)) {} }",
+             R"(
+stdin:1:24-35: ERROR: has_key() only accepts maps that have keys to iterate over
```

Comment:
"maps that have keys to iterate over" is odd phrasing - we're not going to iterate over any keys with this function

Reply:
copy 🍝